addons/tls13-middlebox/bootstrap.js

```diff
@@ -0,0 +1,276 @@
+"use strict";
+
+let {classes: Cc, interfaces: Ci, utils: Cu, results: Cr} = Components;
+
+Cu.import("resource://gre/modules/Preferences.jsm");
+Cu.import("resource://gre/modules/TelemetryController.jsm");
+Cu.import("resource://gre/modules/XPCOMUtils.jsm");
+
+const VERSION_MAX_PREF = "security.tls.version.max";
+const FALLBACK_LIMIT_PREF = "security.tls.version.fallback-limit";
+
+let readwrite_prefs = new Preferences({defaultBranch: true});
+
+// all combination of configurations we care about.
+let configurations = [
+  {maxVersion: 4, fallbackLimit: 4, website: "enabled.tls13.com"},
+  {maxVersion: 4, fallbackLimit: 4, website: "disabled.tls13.com"},
+  {maxVersion: 3, fallbackLimit: 3, website: "control.tls12.com"}
+];
+
+const CERT_USAGE_SSL_CLIENT      = 0x0001;
+const CERT_USAGE_SSL_SERVER      = 0x0002;
+const CERT_USAGE_SSL_CA          = 0x0008;
+const CERT_USAGE_EMAIL_SIGNER    = 0x0010;
+const CERT_USAGE_EMAIL_RECIPIENT = 0x0020;
+const CERT_USAGE_OBJECT_SIGNER   = 0x0040;
+
+let certDB = Cc["@mozilla.org/security/x509certdb;1"].getService(Ci.nsIX509CertDB);
+
+// some fields are not available sometimes, so we have to catch the errors and return undefined.
+function getFieldValue(obj, name) {
+  try {
+    return obj[name];
+  } catch (ex) {
+    return undefined;
+  }
+}
+
+// enumerate nsIX509CertList data structure and put elements in the array
+function nsIX509CertListToArray(list) {
+  let array = [];
+
+  let iter = list.getEnumerator();
+
+  while (iter.hasMoreElements()) {
+    array.push(iter.getNext().QueryInterface(Ci.nsIX509Cert));
+  }
+
+  return array;
+}
+
+// veries the cert using either SSL_SERVER or SSL_CA usages and extracts the chain
+// returns null in case an error occurs
+function getCertChain(cert, usage) {
+  return new Promise((resolve, reject) => {
```

Comment:
Sounds like a good idea. I just tried it; however, since the return statements are in the callback, this approach wouldn't work. We need to have a return statement in the getCertChain function itself, which is not possible due to another async call to asyncVerifyCertAtTime.